data/transactions/logic/eval.go

```diff
@@ -425,41 +547,43 @@ func (pe PanicError) Error() string {
 var errLogicSigNotSupported = errors.New("LogicSig not supported")
 var errTooManyArgs = errors.New("LogicSig has too many arguments")
 
-// EvalStatefulCx executes stateful TEAL program
-func EvalStatefulCx(program []byte, params EvalParams) (bool, *EvalContext, error) {
-	var cx EvalContext
-	cx.EvalParams = params
-	cx.runModeFlags = runModeApplication
+// EvalContract executes stateful TEAL program as the gi'th transaction in params
+func EvalContract(program []byte, gi int, aid basics.AppIndex, params *EvalParams) (bool, *EvalContext, error) {
+	if params.Ledger == nil {
+		return false, nil, errors.New("no ledger in contract eval")
+	}
+	cx := EvalContext{
+		EvalParams:   params,
+		runModeFlags: runModeApplication,
+		GroupIndex:   gi,
+		Txn:          &params.TxnGroup[gi],
+		appID:        aid,
+	}
 	pass, err := eval(program, &cx)
 
-	// The following two updates show a need for something like a
-	// GroupEvalContext, as we are currently tucking things into the
-	// EvalParams so that they are available to later calls.
-
-	// update pooled budget
-	if cx.Proto.EnableAppCostPooling && cx.PooledApplicationBudget != nil {
-		// if eval passes, then budget is always greater than cost, so should not have underflow
-		*cx.PooledApplicationBudget = basics.SubSaturate(*cx.PooledApplicationBudget, uint64(cx.cost))
-	}
 	// update side effects
-	cx.PastSideEffects[cx.GroupIndex].setScratchSpace(cx.scratch)
+	cx.pastScratch[cx.GroupIndex] = &scratchSpace{}
+	*cx.pastScratch[cx.GroupIndex] = cx.scratch
```

Comment:
pastScratch holds all the scratch values from previous executions.  This stashes the current cx.scratch in there for future txn use.

Reply:
Oh, sorry, you mean why not store the pointer.  I was tempted.  I thought it might have bad garbage collection effects if I held onto a pointer into the middle of the cx, which is what that would be.  In many GCs, that would mean the entire cx is held onto.  And it also felt nice and clean to copy the values.  But I could be convinced to change.

Reply:
I think you are right. Please then add a comment :)